### PR TITLE
Disable settings in the back end if defined in the container

### PR DESCRIPTION
**Pull request by dmolineus**
If any settings of the `localconfig` are defined by the `contao.localconfig` parameter the settings made in the Contao backend are ignored. 

However it's possible to edit the settings in the backend, but the changes are not visible after save+reload. This PR disables all settings which are defined by parameter so that you easily see that these settings are overriden by `contao.localconfig`. Furthermore the `important.svg` icon is added to the label to explain why the setting is disabled.

This might be helpful if you work with different administrators (where some don't have access to the code). I decided not to remove the fields, so the values are visible and people won't get confused about the missing configurations.

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

**Comment by aschempp** [comment]
Great addition! I would actually consider this a bugfix, because it fixes the case that a setting is not applied (and I'd like to have this in 4.9 😏 )

**Review by aschempp — DISMISSED** [review]

**Review by fritzmg — DISMISSED** [review]

**Review by aschempp — COMMENTED** [review]
can we not inject the parameter directly?

**Comment by dmolineus** [comment]
> can we not inject the parameter directly?

This would lead to an error if the parameter is not defined. Of course I could remove the listener then instead

**Comment by aschempp** [comment]
Maybe we should make sure it's always defined in https://github.com/contao/contao/blob/master/core-bundle/src/DependencyInjection/ContaoCoreExtension.php#L82 🤷‍♂

**Comment by dmolineus** [comment]
> Maybe we should make sure it's always defined in https://github.com/contao/contao/blob/master/core-bundle/src/DependencyInjection/ContaoCoreExtension.php#L82 man_shrugging

I reworked the listener so that `contao.localconfig` parameter is injected directly.

**Review by aschempp — CHANGES_REQUESTED** [review]
I really like this! ❤️ 

One thing: calling this _parameters_ is not correct, this is the bundle configuration. We're only "storing" it in parameters to make it available after compilation. How about changing the error message to something like "Setting is overridden by bundle configuration"? And probably also rename the listener class?

**Comment by dmolineus** [comment]
> I really like this! 
> 
> One thing: calling this _parameters_ is not correct, this is the bundle configuration. We're only "storing" it in parameters to make it available after compilation. How about changing the error message to something like "Setting is overridden by bundle configuration"? And probably also rename the listener class?

I renamed it to configured by bundle / bundle configuration.

File: core-bundle/src/Resources/contao/languages/en/tl_settings.xlf [hunk 1]
```diff
@@ -167,11 +167,11 @@
...
       </trans-unit>
-      <trans-unit id="tl_settings.parameterBasedSetting.0">
-        <source>Parameter based setting</source>
+      <trans-unit id="tl_settings.configuredInBundle.0">
+        <source>Configured in bundle configuration</source>
       </trans-unit>
-      <trans-unit id="tl_settings.parameterBasedSetting.1">
-        <source>Setting is defined by contao.localconfig parameter.</source>
+      <trans-unit id="tl_settings.configuredInBundle.1">
+        <source>Setting is defined by contao core bundle configuration.</source>
```
Comment (xchs) [inline core-bundle/src/Resources/contao/languages/en/tl_settings.xlf]:
```suggestion
        <source>Setting is defined by Contao core bundle configuration.</source>
```

**Review by aschempp — DISMISSED** [review]
Great work, thank you!

File: phpstan.neon.dist [hunk 2]
```diff
@@ -19,7 +19,7 @@ parameters:
         - '#Constant TL_SCRIPT not found\.#'
 
         # Ignore the Adapter::missingMethod() warning
-        - '#Call to an undefined method Contao\\CoreBundle\\Framework\\Adapter::missingMethod\(\)\.#'
+        - '#Call to an undefined method Contao\\CoreBundle\\Framework\\Adapter::.+\(\)\.#'
```
Comment (leofeyer) [inline phpstan.neon.dist]: Why did you change this?

Comment (dmolineus) [reply phpstan.neon.dist]: I use the adapter to decorate the `Contao\Image::getHtml` method, see https://github.com/contao/contao/pull/1382/files#diff-5ea56ee23b04ee35ebcf7c9d0e8b3df0R63

I got an undefined method warning. I adjusted it so that all missing method warnings would be silenced.

Comment (leofeyer) [reply phpstan.neon.dist]: But that is not the correct fix, I'm afraid. Instead, you should add an inline phpDoc comment:

```php
        /** @var Image $adapter */
        $adapter = $this->framework->getAdapter(Image::class);

        return $adapter->getHtml('…');
```

Comment (m-vo) [reply phpstan.neon.dist]: Imo the doc comments should also include the adapter:
```php
 /** @var Image&Adapter $adapter */`
```

Comment (leofeyer) [reply phpstan.neon.dist]: No, that is wrong. The object is not an actual instance of the `Image` class, therefore – if you really want to add the `Adapter` class – it would be `Image|Adapter`. We are not adding the `Adapter` class though.

Comment (m-vo) [reply phpstan.neon.dist]: :thinking: Imo omitting `Adapter` will potentially induce problems: You're relying on the static analysis tools to ignore the return type of `$framework->getAdapter()` in this case. I'm not sure if this would even work with higher phpstan levels?

Widening the type is imo the correct approach here (in the sense of duck typing like with a mock object). It won't ever be an `Image` (therefore `Image|Adapter` is wrong)  but it will have all it's properties.

Also, have a look at these examples:
```php
// this will break (invalid return type)
function getImageAdapter(): Image {
    return $this->framework->getAdapter(Image::class);
}

// this won't break but isn't helpful for the consumer (missing type information)
function getImageAdapter(): Adapter {
    return $this->framework->getAdapter(Image::class);
}

// this is typesafe and helpful :-)
/**
* @return Adapter&Image
*/
function getImageAdapter(): Adapter {
    return $this->framework->getAdapter(Image::class);
}
```

Comment (leofeyer) [reply phpstan.neon.dist]: Using an intersection type is still wrong. 🤷‍♂️ 

> `Foo&Bar` means that the type of a variable is Foo and Bar at the same time.

This condition is clearly not met here!

Comment (dmolineus) [reply phpstan.neon.dist]: Neighter `\Image` nor `\Image&Adapter` is correct here imho. Using phpstan @mixin should be the way to go. I update the PR accordingly. https://phpstan.org/writing-php-code/phpdocs-basics#mixins

Example:

```php
<?php declare(strict_types = 1);

class HelloWorld
{
	public static function sayHello(DateTimeImmutable $date): void
	{
		echo 'Hello, ' . $date->format('j. n. Y');
	}
}

/** 
 * @template T 
 * @mixin T
 */
class Shouter
{
	/** 
	 * @var class-string<T> 
	 */
	private $className;
	
	/** 
	 * @param class-string<T> $className
	 */
	public function __construct(string $className)
	{
		$this->className = $className;
	}
	
	/** 
	 * @param mixed[] $arguments
	 * @return mixed 
	 */
	public function __calc(string $method, array $arguments)
	{
		return $this->className::$method(... $arguments);
	}
}

$shouter = new Shouter(HelloWorld::class);
$shouter->sayHello(new DateTimeImmutable());
```

Comment (dmolineus) [reply phpstan.neon.dist]: Resolved by annotating it with the `@mixin` annotation which works pretty well. See ef2cf67.

Comment (m-vo) [reply phpstan.neon.dist]: >> Foo&Bar means that the type of a variable is Foo and Bar at the same time.
>
>This condition is clearly not met here!

Of course, it is? It's duck typing. Otherwise you could only ever use Interfaces in intersection types.

Comment (leofeyer) [reply phpstan.neon.dist]: @dmolineus You are absolutely right. I have encapsulated your changes into a separate PR for Contao 4.9 in #1836.

File: core-bundle/src/Framework/Adapter.php [hunk 3]
```diff
@@ -20,6 +20,8 @@
  * @mixin T
  *
  * @internal Do not use this class in your code; use ContaoFramework::getAdapter() instead
+ * @template T
+ * @mixin T
```
Comment (leofeyer) [inline core-bundle/src/Framework/Adapter.php]: This is redundant now and needs to be removed. 🙈

File: core-bundle/src/Resources/config/listener.yml [hunk 4]
```diff
@@ -273,3 +273,11 @@ services:
...
         tags:
             - kernel.event_listener
+
+    Contao\CoreBundle\EventListener\DataContainer\DisableBundleConfiguredSettingsListener:
+        public: true
+        arguments:
+            - '@translator'
+            - '@contao.framework'
+        tags:
+            - { name: contao.callback, table: tl_settings, target: config.onload }
```
Comment (leofeyer) [inline core-bundle/src/Resources/config/listener.yml]: You should use annotations instead (see https://github.com/contao/contao/pull/1762/files):

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Framework/Adapter.php (hunk 3 above)
Comment (dmolineus) [reply core-bundle/src/Framework/Adapter.php]: Looks like someone didn't adjust in on merge. :-D Resolved with 43e9d1d

File: core-bundle/src/Resources/config/listener.yml (hunk 4 above)
Comment (dmolineus) [reply core-bundle/src/Resources/config/listener.yml]: Okay, I changed the way of configuration. I honestly how the unit tests should look like to verify the correct implementation and the proposed PR doesn't test it at all. So I at least check if the annotation is there and the interface is implemented in 

https://github.com/contao/contao/pull/1382/commits/59e4b18039cd293c21662ff043569999ecd4507e#diff-5eec131ad59b19ad8080c1eb5e46225fR26-R42

**Comment by dmolineus** [comment]
Failing unit test is not related to this changes.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
@dmolineus Here is what I have changed so you can review:

1. Replaced `ContaoFrameworkInterface` with `ContaoFramework`, because the interface is deprecated.
2. Add `- '%contao.localconfig%'` as third service argument to get rid of the missing arguments exception.
3. Use the info icon instead of the error icon to clarify that the disabled field is not an error.

File: core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php [hunk 5]
```diff
@@ -17,16 +17,13 @@
...
-use PHPUnit\Framework\MockObject\MockObject;
 use Symfony\Contracts\Translation\TranslatorInterface;
-use Terminal42\ServiceAnnotationBundle\ServiceAnnotationInterface;
 
 class DisableBundleConfiguredSettingsListenerTest extends TestCase
 {
     public function testAnnotatedCallbacks(): void
     {
         $listener = $this->createListener();
-        $this->assertInstanceOf(ServiceAnnotationInterface::class, $listener);
```
Comment (dmolineus) [inline core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php]: Why did you remove it? Shouldn't it be ensured that the interface is implemented otherwise the auto configuration won't work.

**Review by leofeyer — CHANGES_REQUESTED** [review]
I have noticed that you have originally named the class `DisableParametersBasedSettingsListener` and later changed the name to `BundleConfigured` and `configuredInBundle` in 3eeed85a018c60cf782d4bc229d48ec754544292.  IMHO the renaming is wrong, because `contao.localconfig` is not defined in a bundle configuration but in the app configuration.

What about `DisableAppConfiguredSettingsListener` or `DisableContainerConfiguredSettingsListener` instead?

File: core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php [hunk 6]
```diff
@@ -17,16 +17,13 @@
...
 
         $annotationReader = new AnnotationReader();
         $annotation = $annotationReader->getClassAnnotation(new \ReflectionClass($listener), Callback::class);
```
Comment (dmolineus) [inline core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php]: If you won't readd `$this->assertInstanceOf(ServiceAnnotationInterface::class, $listener);` there's no need for an instance of the listener. Then the class string `DisableBundleConfiguredSettingsListener::class` can be used for the reflection.

**Review by dmolineus — COMMENTED** [review]
LGTM besides a question of the adjusted unit test.

File: core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php (hunk 5 above)
Comment (leofeyer) [reply core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php]: PhpStorm told me that the assertion can be skipped:

**Comment by dmolineus** [comment]
> What about DisableAppConfiguredSettingsListener or DisableContainerConfiguredSettingsListener instead?

Agreed.

File: core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php (hunk 5 above)
Comment (dmolineus) [reply core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php]: Imho the annotation of EA is wrong as it does not know that it's a marker interface for the magic auto configuration.

Comment (leofeyer) [reply core-bundle/tests/EventListener/DataContainer/DisableBundleConfiguredSettingsListenerTest.php]: Agreed. I have re-added it in cde293e2137787a37497a3184edd32043b9213f2.

**Comment by leofeyer** [comment]
> > What about DisableAppConfiguredSettingsListener or DisableContainerConfiguredSettingsListener instead?
> 
> Agreed.

Which one? 😄

**Comment by dmolineus** [comment]
> > > What about DisableAppConfiguredSettingsListener or DisableContainerConfiguredSettingsListener instead?
> > 
> > 
> > Agreed.
> 
> Which one? smile

I would use `DisableAppConfiguredSettingsListener`. Do you change it or shall I (won't happen before the weekend on my side).

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
> I would use `DisableAppConfiguredSettingsListener`. Do you change it or shall I (won't happen before the weekend on my side).

Changed in 7ba2c65ffc0c17183d9cff82801e66a23204ae9d.

**Comment by leofeyer** [comment]
Thank you @dmolineus.
